manager-parent/manager-v2/src/main/java/com/redhat/service/smartevents/manager/v2/metrics/MetricsServiceImpl.java

```diff
@@ -0,0 +1,107 @@
+    String operationDurationMetricName;
+
+    @Override
+    public Set<String> getMetricNames() {
+        return Set.of(operationDurationMetricName, operationTotalCountMetricName, operationTotalSuccessCountMetricName, operationTotalFailureCountMetricName);
+    }
+
+    @Override
+    public void onOperationStart(ManagedResourceV2 managedResource, MetricsOperation operation) {
+        incrementCounter(operationTotalCountMetricName, buildTags(managedResource, operation));
+    }
+
+    private void incrementCounter(String counterName, List<Tag> tags) {
+        meterRegistry.counter(counterName, tags).increment();
+    }
+
+    @Override
+    public void onOperationComplete(ManagedResourceV2 managedResource, MetricsOperation operation) {
+        if (isOperationSuccessful(managedResource, operation)) {
+            incrementCounter(operationTotalSuccessCountMetricName, buildTags(managedResource, operation));
+            recordOperationDuration(managedResource, operation);
+        }
+    }
+
+    private boolean isOperationSuccessful(ManagedResourceV2 managedResource, MetricsOperation operation) {
+        ManagedResourceStatus status = StatusUtilities.getManagedResourceStatus(managedResource);
+        if (MetricsOperation.MANAGER_RESOURCE_PROVISION == operation || MetricsOperation.MANAGER_RESOURCE_MODIFY == operation) {
+            return ManagedResourceStatusV2.READY == status;
+        }
+
+        return ManagedResourceStatusV2.DELETED == status;
+    }
+
+    @Override
+    public void onOperationFailed(ManagedResourceV2 managedResource, MetricsOperation operation) {
+        if (isOperationFailed(managedResource)) {
+            incrementCounter(operationTotalFailureCountMetricName, buildTags(managedResource, operation));
+            recordOperationDuration(managedResource, operation);
+        }
+    }
+
+    private boolean isOperationFailed(ManagedResourceV2 managedResource) {
+        ManagedResourceStatus status = StatusUtilities.getManagedResourceStatus(managedResource);
+        return ManagedResourceStatusV2.FAILED == status;
+    }
+
+    private List<Tag> buildTags(ManagedResourceV2 managedResource, MetricsOperation operation) {
+        Tag instanceTag = Tag.of(RESOURCE_TAG, managedResource.getClass().getSimpleName().toLowerCase());
+        Tag versionTag = Tag.of(VERSION_TAG, V2.class.getSimpleName());
+        return List.of(instanceTag, versionTag, operation.getMetricTag());
+    }
+
+    private void recordOperationDuration(ManagedResourceV2 managedResource, MetricsOperation operation) {
+        Duration operationDuration = calculateOperationDuration(managedResource.getOperation());
+        meterRegistry.timer(operationDurationMetricName, buildTags(managedResource, operation)).record(operationDuration);
+    }
+
+    private Duration calculateOperationDuration(Operation operation) {
+        return Duration.between(operation.getRequestedAt(), ZonedDateTime.now(ZoneOffset.UTC));
```

Comment:
@manstis Shouldn't we be persisting the time the operation was completed on the `Operation` resource? We set it when the user first requests a create/update/delete, so shouldn't we additionally record the timestamp when we first determine the operation to be either failed or completed?

Reply:
The "completion" timestamp is never used other than to [calculate the duration](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/main/manager-parent/manager-v1/src/main/java/com/redhat/service/smartevents/manager/v1/metrics/MetricsServiceImpl.java#L91-L102) metric.

This is why we decided not to store it in `v2`.